sdk/core/Azure.Core/src/ETag.cs

```diff
@@ -83,17 +83,20 @@ internal static ETag Parse(string value)
             {
                 return All;
             }
-            else if (value.StartsWith("W/", StringComparison.Ordinal))
+            else if (!(value.StartsWith(QuoteString, StringComparison.Ordinal) || value.StartsWith(WeakETagPrefix, StringComparison.Ordinal)) ||
+                 !value.EndsWith(QuoteString, StringComparison.Ordinal))
             {
-                throw new NotSupportedException("Weak ETags are not supported.");
+                throw new ArgumentException("The value should be equal to * , be wrapped in quotes, or be wrapped in quotes prefixed by W/", nameof(value));
             }
-            else if (!value.StartsWith(QuoteString, StringComparison.Ordinal) ||
-                     !value.EndsWith(QuoteString, StringComparison.Ordinal))
+
+            if (value.StartsWith(WeakETagPrefix, StringComparison.Ordinal))
             {
-                throw new ArgumentException("The value should be equal to * or be wrapped in quotes", nameof(value));
+                return new ETag(value);
```

Comment:
Are we sure this is the right representation?

Should we have IsWeak property and parse the value out? Because we do trim quotes from the normal etag.

Reply:
If we did parse down to just the string inside the quotes, wouldn't we need to re-wrap it on `ToString()`? If so, what is the purpose of storing the naked string?

Reply:
@KrzysztofCwalina I think you were interested to be involved in the change to support weak ETag.

Reply:
https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core/src/Shared/ConditionalRequestOptionsExtensions.cs#L10-L40 FYI

Reply:
I'm probably missing some background or context here, but would it be simpler and more consistent if we always stored the actual value of the `ETag` regardless of type and eliminated any transformations in or out? We already store the raw value if the `ctor` is used, for example.

Reply:
It could've been :) but that ship has sailed. Now we need to think about this type being backwards compatible.

It would be nice to figure out all the design holes in it though.

Reply:
As discussed offline, we are leaning towards 
- keeping the behavior of returning the raw value for weak etags 
- implementing a new method that returns the fully formatted (wrapped with quotes or weak etag prefix) for either type. 
- hiding the current `ToString` override and never use the override in our code.

Reply:
https://github.com/dotnet/roslyn-analyzers#microsoftcodeanalysisbannedapianalyzers might be useful to prevent anyone from using ToString in the future.

Reply:
I've started a separate branch to implement the new behavior, since it affects many other clients. What do we need to do in terms of sequence of events so that the change to only use the new method gets picked up by generated code as well?

Reply:
Because of the need to ship Core before updating generation, this PR will contain the ETag changes after all.